hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/EnrichmentProcessor.java

```diff
@@ -44,7 +58,10 @@ public void process(StructuredTrace trace) {
     AvroToJsonLogger.log(LOG, "Structured Trace before all the enrichment is: {}", trace);
     for (Enricher enricher : enrichers) {
       try {
-        applyEnricher(enricher, trace);
+        simpleTimeLimiter.callWithTimeout(
+            () -> applyEnricher(enricher, trace),
```

Comment:
As we have a dag of enrichers (most of them are connected and ordered), I think, a small number of parallel threads will be fine.

So, in the worst case, the config to avoid any rebalancing (with this change):
- heartbeat.interval.ms = 10s * number of enrichers * (ms multiplier)
- session.timeout.ms = 3 * heartbeat.interval.ms

This change will help us as a safety net at the cost of an unknown state of downstream enrichment from the timeout enricher. Is this correct? 

We can move with this change. But, I think, we should derive the upper bound (session + heartbeat)  from metrics. (Total time taken by all enrichers on maxTraceSize). 

Later, some of the other ways, we should handle enrichment time that depends on (as the above knob is also manual adjustment):
- number of enrichers (we can divide them into multiple enrichment processes writing to its own Kafka topics, and creating a pipe as per dag)
- maxTraceSize (we already have control for this)
- enrichment logic (we should somehow cap this while adding a new one by dividing into multiple if possible and accordingly adjusting above heartbeat and session time out)

Reply:
> This change will help us as a safety net at the cost of an unknown state of downstream enrichment from the timeout enricher. Is this correct?
That is correct

> We can move with this change. But, I think, we should derive the upper bound (session + heartbeat) from metrics. (Total time taken by all enrichers on maxTraceSize).
This is worth checking for. However max is probably a wrong number of pick. Probably between 95-99 percentile, if that data is available

>  number of enrichers (we can divide them into multiple enrichment processes writing to its own Kafka topics, and creating a pipe as per dag)
Internal pipe also reflects as kafka topics. Not sure how flexible is that.

Reply:
> This is worth checking for. However max is probably a wrong number to pick. Probably between 95-99 percentile, if that data is available

we do have metrics for this measuring enrichment process - 
https://github.com/hypertrace/hypertrace-ingester/blob/main/hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java#L42

Reply:
> Internal pipe also reflects as kafka topics. Not sure how flexible is that.

This is not currently configured. So, we should explore this option too. However, I am not fully sure. In my current understanding, the entire topology is processed into the same poll cycle of the source topics of a stream including punctuators.